packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -536,25 +539,69 @@ export default class CodeGen {
 
         const args: t.Expression[] = [t.callExpression(identifier, []), t.literal(key)];
 
-        // Only add the third argument (databag) if this element needs it
-        if (element.listeners.length || element.directives.length) {
-            const databagProperties: t.Property[] = [];
+        // Only add the third argument (staticParts) if this element needs it
+        const staticParts = this.genStaticParts(element);
+        if (staticParts) {
+            args.push(staticParts);
+        }
+
+        return this._renderApiCall(RENDER_APIS.staticFragment, args);
+    }
 
-            // has event listeners
-            if (element.listeners.length) {
-                databagProperties.push(this.genEventListeners(element.listeners));
+    genStaticParts(element: StaticElement): t.ArrayExpression | undefined {
+        const stack: (StaticElement | Text)[] = [element];
+        const partIdsToDatabagProps = new Map<number, t.Property[]>();
+        let partId = -1;
+
+        const addDatabagProp = (prop: t.Property) => {
+            let databags = partIdsToDatabagProps.get(partId);
+            if (!databags) {
+                databags = [];
+                partIdsToDatabagProps.set(partId, databags);
             }
+            databags.push(prop);
+        };
+
+        // Depth-first traversal. We assign a partId to each element, which is an integer based on traversal order.
+        while (stack.length > 0) {
+            const node = stack.shift()!;
+            partId++;
+            if (isElement(node)) {
+                // has event listeners
+                if (node.listeners.length) {
+                    addDatabagProp(this.genEventListeners(node.listeners));
+                }
 
-            // see STATIC_SAFE_DIRECTIVES for what's allowed here
-            for (const directive of element.directives) {
-                if (directive.name === 'Ref') {
-                    databagProperties.push(this.genRef(directive));
+                // see STATIC_SAFE_DIRECTIVES for what's allowed here
+                for (const directive of node.directives) {
+                    if (directive.name === 'Ref') {
+                        addDatabagProp(this.genRef(directive));
+                    }
+                }
+
+                // For depth-first traversal, prepend to the stack in reverse order
+                for (let i = node.children.length - 1; i >= 0; i--) {
+                    const childNode = node.children[i];
+                    stack.unshift(childNode);
```

Comment:
This does mirror the diffing algo, yes – depth-first tree-traversal order.

The problem with `push` and `pop` is that they would append to the end of the array. We need to process all children in order before processing any siblings (aunts/uncles).

Probably this code would be clearer:

```js
stack.push(...node.children)
```

I think this may actually have roughly the same perf characteristics. I'll make that change.

Reply:
0b78dd440